lib/pbench/test/unit/server/test_templates.py

```diff
@@ -0,0 +1,146 @@
+import datetime
+import io
+from pathlib import Path
+import pytest
+
+from pbench.server.templates import JsonFile, JsonToolFile
+
+
+@pytest.fixture()
+def fake_resolve(monkeypatch):
+    """
+    JsonFile objects use Path().resolve(strict=True) to get a full path; fake
+    it so we don't need real files.
+
+    Args:
+        monkeypatch: Patching fixture
+    """
+
+    def fake_resolve(self: Path, strict: bool = False):
+        """
+        Return a "full" fake file name
+
+        Args:
+            file: filename
+
+        Returns:
+            new Path with fake resolution
+        """
+        return Path("/fake/path/") / self.name
+
+    with monkeypatch.context() as m:
+        m.setattr(Path, "resolve", fake_resolve)
+        yield
+
+
+@pytest.fixture()
+def fake_open(monkeypatch, request):
+    """
+    JsonFile objects use Path().resolve(mode="r") to read the JSON file; fake
+    it so we don't need real files.
+
+    Args:
+        monkeypatch: Patching fixture
+    """
+    data = request.node.get_closest_marker("open_data")
+    if data is None:
+        data = "{'default': 'test'}"
+
+    def fake_open(self: Path, mode: str = "r"):
+        """
+        Return an IO stream
+
+        Args:
+            file: filename
+
+        Returns:
+            IO channel from string
+        """
+        return io.StringIO(data)
+
+    with monkeypatch.context() as m:
+        m.setattr(Path, "open", fake_open)
+        yield
+
+
+class TestJsonFile:
+    def test_json_file(self, fake_resolve, fake_mtime):
+        json = JsonFile(Path("mapping.json"))
+        assert str(json.file) == "/fake/path/mapping.json"
+        assert json.json is None
+        assert json.modified == datetime.datetime(2021, 1, 29, 0, 0, 0)
+
+    def test_json_load(self, monkeypatch, fake_resolve, fake_mtime):
+        json = JsonFile(Path("mapping.json"))
+        with monkeypatch.context() as m:
+
+            def fake_open(self: Path, mode: str = "r"):
+                return io.StringIO('{"_meta": {"version": 6}}\n')
+
+            m.setattr(Path, "open", fake_open)
+            # m.setattr(Path, "open", lambda self, mode : io.StringIO('{"_meta": {"version": 6}}\n'))
```

Comment:
Remove this line before merging?  Ditto line 109, et al.

(A lambda seems like a nice idea, here...you couldn't get it to work?)

Reply:
Yeah, couldn't get it to work. I gave up. And the problem with having posted this with staged commits is that you keep commenting on stuff in earlier commits that don't exist anymore. I think my strategy was flawed. ;-)

Reply:
I think your strategy was good, but, perhaps, your execution contained a few flaws.  :-)